ir/src/boundary_constraints.rs

```diff
@@ -95,6 +95,40 @@ impl BoundaryConstraints {
                 },
                 _ => unimplemented!(),
             },
+            IdentifierType::TraceColumnGroup(column_group) => {
+                if let TraceColAccess::GroupAccess(name, idx) = constraint.column() {
+                    if *idx >= column_group.size() {
+                        return Err(SemanticError::vector_access_out_of_bounds(
+                            &VectorAccess::new(name.clone(), *idx),
+                            column_group.size(),
+                        ));
+                    }
+                    match column_group.trace_segment() {
+                        0 => match constraint.boundary() {
+                            ast::Boundary::First => self
+                                .main_first
+                                .insert(column_group.start_col_idx() + idx, expr),
+                            ast::Boundary::Last => self
+                                .main_last
+                                .insert(column_group.start_col_idx() + idx, expr),
+                        },
+                        1 => match constraint.boundary() {
+                            ast::Boundary::First => self
+                                .aux_first
+                                .insert(column_group.start_col_idx() + idx, expr),
+                            ast::Boundary::Last => self
+                                .aux_last
+                                .insert(column_group.start_col_idx() + idx, expr),
+                        },
+                        _ => unimplemented!(),
+                    }
```

Comment:
You could condense this a bit by defining a variable before the match for `column_group.start_col_idx() + idx`

Reply:
The code has been refactored.